P0619R4 Removing C++17-Deprecated Features

miscco opened the pull request:
# Description
This implements #28 by conditionally removing deprecated features when compiling in C++20 mode following the guidance from P0619R4

# Checklist

Be sure you've read README.md and understand the scope of this repo.

If you're unsure about a box, leave it unchecked. A maintainer will help you.

- [x] Identifiers in product code changes are properly `_Ugly` as per
  https://eel.is/c++draft/lex.name#3.1 or there are no product code changes.
- [ ] The STL builds successfully and all tests have passed (must be manually
  verified by an STL maintainer before automated testing is enabled on GitHub,
  leave this unchecked for initial submission).
- [x] These changes introduce no known ABI breaks (adding members, renaming
  members, adding virtual functions, changing whether a type is an aggregate
  or trivially copyable, etc.).
- [x] These changes were written from scratch using only this repository,
  the C++ Working Draft (including any cited standards), other WG21 papers
  (excluding reference implementations outside of proposed standard wording),
  and LWG issues as reference material. If they were derived from a project
  that's already listed in NOTICE.txt, that's fine, but please mention it.
  If they were derived from any other project (including Boost and libc++,
  which are not yet listed in NOTICE.txt), you *must* mention it here,
  so we can determine whether the license is compatible and what else needs
  to be done.

BillyONeal commented on stl/inc/type_traits:
```diff
@@ -86,29 +87,46 @@ struct _Arg_types<_Ty1, _Ty2> {
     template <class _Ret, class... _Types>                                                                    \
     struct _Is_function<_Ret(_Types..., ...) CV_REF_NOEXCEPT_OPT> { /* no calling conventions for ellipsis */ \
         using _Bool_type = true_type;                                                                         \
         _CXX17_DEPRECATE_ADAPTOR_TYPEDEFS typedef _Ret result_type;                                           \
```
Instead of needing to duplicate this macro spam I think you could just make the typedef have an `_Ugly` name in 20 like this:

```
#if _HAS_CXX20
#define _RESULT_TYPE_NAME _Unnamable
#else
#define _RESULT_TYPE_NAME result_type
#endif // _HAS_CXX20
```

then name the typedef with that.

CaseyCarter commented:
> I suspect we will need escape hatch macros for these, at least:
> 
> * an escape hatch that brings back the extra typedefs
> * an escape hatch that brings back shared_ptr::unique
> * an escape hatch that brings back is_pod
> * an escape hatch that brings back result_of

I'd prefer to see a single escape hatch in the style of `_HAS_AUTO_PTR_ETC`.

BillyONeal commented:
> > I suspect we will need escape hatch macros for these, at least:
> > 
> > * an escape hatch that brings back the extra typedefs
> > * an escape hatch that brings back shared_ptr::unique
> > * an escape hatch that brings back is_pod
> > * an escape hatch that brings back result_of
> 
> I'd prefer to see a single escape hatch in the style of `_HAS_AUTO_PTR_ETC`.

Our experience with `_HAS_AUTO_PTR_ETC` is precisely why that scares me -- looking at the number of RWC projects that need that does not fill me with confidence that such a granular switch will ever be disabled by large projects.

Then again I'm on record saying the standard shouldn't ever remove anything :)

barcharcraz commented:
if we continue to deprecate and remove stuff from the standard but not the code it may make sense to have something like `_CXX17_DEPRECATE_ADAPTOR_TYPEDEFS ` and friends that hard errors on their use (or removes them from the headers).

StephanTLavavej commented:
Having proposed WG21-N4190 and implemented `_HAS_AUTO_PTR_ETC`, I now believe that having a single mega-escape-hatch is counterproductive. That is, it should have been `_HAS_AUTO_PTR`, `_HAS_RANDOM_SHUFFLE`, and `_HAS_OLD_FUNCTIONAL_STUFF`. Note that we have fine-grained escape hatches for the C++17 removals P0003R5 Removing Dynamic Exception Specifications (`_HAS_UNEXPECTED`) and P0004R1 Removing Deprecated Iostreams Aliases (`_HAS_OLD_IOSTREAMS_MEMBERS`), which (happily!) are virtually never activated because those things are obscure.

I think that following the design we've been using for deprecations would be optimal - that is, fine-grained escape hatches, which are also controlled by an overall escape hatch. The naming for the overall escape hatch could be verbose like `_HAS_FEATURES_REMOVED_BY_CXX20`.

(Separately, we could retrofit the C++17 removal macros accordingly, but I'd want to see that as a followup change, and no action might be necessary there.)

CaseyCarter commented:
>  I now believe that having a single mega-escape-hatch is counterproductive

I think having only a coarse-grained hatch encourages faster migration. Folks who don't care about deprecation/removal also don't care about coarse vs. fine - they'll just hit the big red button and go about their business either way. Folks who *do* care about avoiding dependencies on deprecated/removed features will eradicate all uses from local code, and only use escape hatches for third-party code. If we give them fine-grained hatches, they will enable only those necessary for their third-party code to compile thereby reducing their exposure. If we give them only coarse-grained hatches, they will have to accept maximum exposure to use any third-party code that uses a deprecated/removed feature. Being more dissatisfied in the latter case encourages the motivated users to push harder on unmotivated authors of third-party code.

That said, I am motivated just enough by this question to type this explanation - I won't die on this hill, but I'll yell loudly as I retreat ;)

StephanTLavavej commented:
We still have to enable `_HAS_AUTO_PTR_ETC` for many Real World Code projects, despite the size of the option. I understand your argument, but it seems unlikely that "motivated users" would notice the size of a coarse-grained escape hatch - how many users realized what `ETC` controlled?

My argument is straightforward - we use a fine+coarse combo for deprecations, following that for removals seems consistent.

miscco commented:
I started implementing the changes, but for adaptor typedefs it is a bit more work.

miscco commented:
I added escape hatches to retain the removed features. I hope that this is what was expected

BillyONeal commented on stl/inc/ostream:
```diff
@@ -109,7 +109,11 @@ public:
         _STL_DISABLE_DEPRECATED_WARNING
         __CLR_OR_THIS_CALL ~sentry() noexcept {
 #if _HAS_EXCEPTIONS
+#if _HAS_DEPRECATED_UNCAUGHT_EXCEPTION
```
This is still blocked by ArchivedOS-12000909. @stwish-msft Do you know what the status of that is?

miscco replied:
Any update here?

BillyONeal commented on stl/inc/xmemory:
```diff
@@ -804,6 +804,7 @@ public:
     }
 };
 
+#if !_HAS_CXX20
```
The `allocator<void>` specialization was removed from the standard only because the primary template was able to do all the same things. Given that we have escape hatches, I think this needs to always stay.

miscco replied:
I cannot follow. If it has no purpose why should it stay?

BillyONeal replied:
It has a purpose: if the user has used one of the escape hatches to put the typedefs back on the primary template, then it would make `std::allocator<void>` ill formed. The only reason for this specialization to exist is to make naming that type well formed; if we always leave this here, that will always be well formed. Even if the primary template could be used, less conditional compilation is better.

miscco replied:
Kept it for good

AdamBucior commented:
But is_pod hasn't been removed yet. Shouldn't it at least be available by default?

StephanTLavavej commented on stl/inc/type_traits:
```diff
@@ -521,12 +521,14 @@ _INLINE_VAR constexpr bool is_volatile_v<volatile _Ty> = true;
 template <class _Ty>
 struct is_volatile : bool_constant<is_volatile_v<_Ty>> {};
 
+#if !_HAS_DEPRECATED_TYPETRAITS
```
As @AdamBucior observed in https://github.com/microsoft/STL/pull/380#issuecomment-566749706 , `is_pod` wasn't removed from C++20.

miscco replied:
Done

StephanTLavavej commented on stl/inc/type_traits:
```diff
@@ -563,6 +565,7 @@ struct is_standard_layout : bool_constant<__is_standard_layout(_Ty)> {}; // dete
 template <class _Ty>
 _INLINE_VAR constexpr bool is_standard_layout_v = __is_standard_layout(_Ty);
 
+#if !_HAS_CXX20
```
Why is `is_literal_type` controlled by `_HAS_CXX20` instead of the `_HAS_DEPRECATED_TYPETRAITS` escape hatch?

miscco replied:
I think i missed it. Fixed

StephanTLavavej commented on stl/inc/type_traits:
```diff
@@ -1695,6 +1699,7 @@ struct _Invoke_traits<void_t<decltype(_STD invoke(_STD declval<_Types>()...))>,
         conjunction_v<_Is_nothrow_invocable, disjunction<is_void<_Rx>, _Is_nothrow_convertible<type, _Rx>>>>;
 };
 
+#if !_HAS_DEPRECATED_TYPETRAITS
```
You're testing `#if !_HAS_DEPRECATED_TYPETRAITS` but it seems to me that this is reversed. After looking at the definition (not just the naming), I believe you intended to guard this machinery with `#if _HAS_DEPRECATED_TYPETRAITS`. Am I completely confused?

miscco replied:
Nope that was me who missed the double negation. fixed

StephanTLavavej commented on stl/inc/exception:
```diff
@@ -20,7 +20,9 @@ _STL_DISABLE_CLANG_WARNINGS
 
+#if _HAS_DEPRECATED_UNCAUGHT_EXCEPTION
 _CXX17_DEPRECATE_UNCAUGHT_EXCEPTION _CRTIMP2_PURE bool __CLRCALL_PURE_OR_CDECL uncaught_exception() noexcept;
+#endif //_HAS_DEPRECATED_UNCAUGHT_EXCEPTION
```
Style: We put a space after `//` comments.

miscco replied:
Yep

StephanTLavavej commented on stl/inc/functional:
```diff
@@ -314,6 +314,7 @@ struct bit_not<void> {
     }
 };
 
+#if !_HAS_DEPRECATED_NEGATE
```
Is this test reversed? (I suspect it was originally `#if !_HAS_CXX20` which would explain how this happened.)

miscco replied:
It indeed is

StephanTLavavej commented on stl/inc/memory:
```diff
@@ -417,7 +417,7 @@ _NoThrowFwdIt uninitialized_value_construct_n(_NoThrowFwdIt _First, const _Diff
 #endif // _HAS_CXX17
 
-
+#if !_HAS_CXX20
```
`raw_storage_iterator` should probably be controlled by the usual system of coarse/fine-grained escape hatches.

miscco replied:
Done under `_HAS_DEPRECATED_RAW_STORAGE_ITERATOR`

StephanTLavavej commented on stl/inc/memory:
```diff
@@ -1205,10 +1206,12 @@ public:
         return get()[_Idx];
     }
 
+#if !_HAS_DEPRECATED_SHARED_PTR_UNIQUE
```
Another reversed test?

miscco replied:
yes

StephanTLavavej commented on stl/inc/type_traits:
```diff
@@ -2258,7 +2264,9 @@ namespace _DEPRECATE_TR1_NAMESPACE tr1 {
     using _STD is_member_object_pointer;
     using _STD is_member_pointer;
     using _STD is_object;
+#if !_HAS_DEPRECATED_TYPETRAITS
```
Another `is_pod` occurrence that shouldn't be guarded.

miscco replied:
true

StephanTLavavej commented on stl/inc/type_traits:
```diff
@@ -2283,7 +2291,9 @@ namespace _DEPRECATE_TR1_NAMESPACE tr1 {
     using _STD cref;
     using _STD ref;
     using _STD reference_wrapper;
+#if !_HAS_DEPRECATED_TYPETRAITS
```
Another reversed test?

miscco replied:
:(

StephanTLavavej commented on stl/inc/xmemory:
```diff
@@ -1908,6 +1910,7 @@ _NoThrowFwdIt _Uninitialized_value_construct_n_unchecked1(_NoThrowFwdIt _UFirst,
 }
 #endif // _HAS_IF_CONSTEXPR
 
+#if !_HAS_CXX20
```
I believe that the temporary buffer machinery should be controlled by coarse/fine-grained escape hatches.

miscco replied:
Done with `_HAS_DEPRECATED_TEMPORARY_BUFFER`

StephanTLavavej commented on stl/src/uncaught_exception.cpp:
```diff
@@ -4,9 +4,9 @@
 #include <exception>
 _STD_BEGIN
-
+#if _HAS_DEPRECATED_UNCAUGHT_EXCEPTION
```
For binary compatibility, `src` must always export `uncaught_exception()` from the DLL, even if user headers decide not to declare it.

Since this definition says `_CRTIMP2_PURE` (which will expand to `__declspec(dllexport)`), I believe that simply not adding guards here is sufficient. If we needed the declaration for some reason, we would have to alter the STL's build system to compile with the fine-grained escape hatch (if not the coarse-grained), but that shouldn't be necessary.

miscco replied:
Removed the guards

StephanTLavavej commented on stl/inc/yvals_core.h:
```diff
@@ -855,6 +855,42 @@
 
 // next warning number: STL4027
 
+// P0619R4 remove deprecated features from C++20
```
I would prefer following #28's spelling of `P0619R4 Removing C++17-Deprecated Features In C++20` here.

miscco replied:
yeah

StephanTLavavej commented on stl/inc/yvals_core.h:
```diff
@@ -855,6 +855,42 @@
+#define _HAS_REMOVED_FEATURES_CXX20 (!_HAS_CXX20)
+#endif // _HAS_REMOVED_FEATURES_CXX20
+
+#ifndef _HAS_DEPRECATED_TYPETRAITS
```
Naming: This should probably say `TYPE_TRAITS` with an underscore.

miscco replied:
Second that

StephanTLavavej commented on stl/inc/yvals_core.h:
```diff
@@ -855,6 +855,42 @@
+#endif // _HAS_DEPRECATED_ADAPTOR_TYPEDEFS
+
+#if !_HAS_DEPRECATED_ADAPTOR_TYPEDEFS
+#define _ARGUMENT_TYPE_NAME        _Unnamable_argument
```
Naming: I recommend `Unnameable`. Merriam-Webster does list the version without an 'e', but says it's seen "less commonly".

miscco replied:
Ok

StephanTLavavej commented on stl/inc/yvals_core.h:
```diff
@@ -855,6 +855,42 @@
+#define _HAS_DEPRECATED_ADAPTOR_TYPEDEFS (!_HAS_CXX20 || _HAS_REMOVED_FEATURES_CXX20)
+#endif // _HAS_DEPRECATED_ADAPTOR_TYPEDEFS
+
+#if !_HAS_DEPRECATED_ADAPTOR_TYPEDEFS
```
Style: We conventionally prefer to avoid negated tests when it's possible. Here, you have an `#if/#else`, so I recommend inverting them.

miscco replied:
Done

StephanTLavavej commented on stl/inc/yvals_core.h:
```diff
@@ -855,6 +855,42 @@
 
 // next warning number: STL4027
 
+// P0619R4 remove deprecated features from C++20
```
This should additionally be listed in a `// _HAS_CXX20 indirectly controls:` section at the top, like `N4190` is.

miscco replied:
Added it after the `// _HAS_CXX20 directly controls:` block

StephanTLavavej commented on stl/inc/yvals_core.h:
```diff
@@ -855,6 +855,42 @@
 // next warning number: STL4027
 
+// P0619R4 remove deprecated features from C++20
+#ifndef _HAS_REMOVED_FEATURES_CXX20
```
Naming: Would `_HAS_FEATURES_REMOVED_BY_CXX20` or `_HAS_FEATURES_REMOVED_IN_CXX20` be clearer?

miscco replied:
True, I went for `_HAS_FEATURES_REMOVED_IN_CXX20`

StephanTLavavej commented on stl/inc/yvals_core.h:
```diff
@@ -855,6 +855,42 @@
+#define _HAS_DEPRECATED_SHARED_PTR_UNIQUE (!_HAS_CXX20 || _HAS_REMOVED_FEATURES_CXX20)
+#endif // _HAS_DEPRECATED_SHARED_PTR_UNIQUE
+
+#ifndef _HAS_DEPRECATED_NEGATE
```
Naming: I recommend `NEGATORS` (following the removed section name) since it's a pair of types and functions. Additionally, `std::negate` is an existing, non-deprecated thing (the operator functor for unary minus).

This is also consistent with `_CXX17_DEPRECATE_NEGATORS` and `_SILENCE_CXX17_NEGATORS_DEPRECATION_WARNING`.

miscco replied:
Yep, fixed

StephanTLavavej commented on stl/inc/yvals_core.h:
```diff
@@ -855,6 +855,42 @@
+#define _HAS_REMOVED_FEATURES_CXX20 (!_HAS_CXX20)
+#endif // _HAS_REMOVED_FEATURES_CXX20
+
+#ifndef _HAS_DEPRECATED_TYPETRAITS
```
Alternatively, I think we should probably have separate escape hatches for `RESULT_OF` and `IS_LITERAL_TYPE`, mirroring the separate deprecation silencing macros. This is because `result_of` will be restored 1000x more than `is_literal_type`.

miscco replied:
I split them up into `_HAS_DEPRECATED_IS_LITERAL_TYPE` and `_HAS_DEPRECATED_RESULT_OF`

miscco commented:
I will try to have a look on it this evening

miscco commented:
I addressed most review comments. As `allocator<void>` was the only one without an escape hat I added one too for good measure.

There are some remaining comments from @BillyONeal that I would like to wait on.

I did not silence the warnings when somebody retains the removed features. They should carry that cost.

Also one stylistic question remains. Currently I use `_HAS_DEPRECATED_FOO`. Should it rather be `_HAS_REMOVED_FOO`?

StephanTLavavej commented:
> I did not silence the warnings when somebody retains the removed features. They should carry that cost.

Agreed! Also consistent with the `auto_ptr` precedent.

> Also one stylistic question remains.

That's a good question - of those two, I prefer `_HAS_REMOVED_MEOW`, because it counteracts the removal, and it makes these macros visually distinct from the deprecation silencing.

However, I worry a little bit about grammatical ambiguity - the intent is "has meow, a removed feature in C++20" (i.e. "removed" is an adjective), but it could also be read as "this macro has removed (verb) meow", which is the opposite sense. The older macros `_HAS_AUTO_PTR_ETC`, `_HAS_UNEXPECTED`, `_HAS_OLD_IOSTREAMS_MEMBERS`, etc. didn't say `REMOVED`, so they weren't affected by such ambiguity.

So, two more ideas:

1. Continue to follow that precedent and simply use `_HAS_MEOW`: so `_HAS_RESULT_OF`, `_HAS_UNCAUGHT_EXCEPTION`, etc. 
2. Use `_HAS_OLD_MEOW` (originally used to describe the "old iostreams members"), thereby indicating that old machinery is being restored. `_HAS_OLD_RESULT_OF`, `_HAS_OLD_UNCAUGHT_EXCEPTION`, `_HAS_OLD_NEGATORS` (I really like that one because they are distinct from the new `not_fn` tech).

The more I think about it, the more I like `_HAS_OLD_MEOW`. It avoids the contradiction in saying that we "have" a "removed" thing, but also indicates that something old and yucky is being requested, so it's mild discouragement.

BillyONeal commented on stl/inc/xmemory:
```diff
@@ -804,6 +804,7 @@ public:
     }
 };
 
+#if _HAS_DEPRECATED_ALLOCATOR_VOID
```
As I indicated before, I don't think `_HAS_DEPRECATED_ALLOCATOR_VOID` should exist since its presence is unobservable.

miscco replied:
Dropped that patch

miscco commented:
I dropped the allocator<void> patch due to @BillyONeal comment

BillyONeal reviewed (approved):
Thanks for your contribution!

This looks good to me provided @StephanTLavavej (the deprecations and removals czar) and @amyw-msft (for  ArchivedOS-12000909 ) are OK with it.

StephanTLavavej commented on stl/inc/ostream:
```diff
@@ -109,7 +109,11 @@ public:
+#if _HAS_DEPRECATED_UNCAUGHT_EXCEPTION
             if (!_STD uncaught_exception()) { // TRANSITION, ArchivedOS-12000909
+#else // ^^^ _HAS_DEPRECATED_UNCAUGHT_EXCEPTION / !_HAS_DEPRECATED_UNCAUGHT_EXCEPTION vvv
+            if (!_STD uncaught_exceptions()) {
```
Style issue: We generally avoid negating integers. I believe that this should be phrased as `if (_STD uncaught_exceptions() == 0)`.

miscco replied:
I switched both to `_STD uncaught_exception{s}() == 0` so that they are consitent

StephanTLavavej replied:
The issue is that the return types aren't consistent - `uncaught_exception()` returns `bool` while `uncaught_exceptions()` returns `int`. Stylistically, we never compare `bool` to `0`, so I'm going to restore that line.

StephanTLavavej commented on stl/inc/yvals_core.h:
```diff
@@ -886,6 +889,54 @@
+#endif // _HAS_FEATURES_REMOVED_IN_CXX20
+
+#ifndef _HAS_DEPRECATED_IS_LITERAL_TYPE
+#define _HAS_DEPRECATED_IS_LITERAL_TYPE (!_HAS_CXX20 || _HAS_FEATURES_REMOVED_IN_CXX20)
```
Hmm, for each of the fine-grained macros, is `!_HAS_CXX20 ||` necessary? This seems overly complicated. I'd recommend simply defaulting to `(_HAS_FEATURES_REMOVED_IN_CXX20)`. It would allow someone in C++14/17 mode to remove all of these features with a single macro, instead of defining each fine-grained macro - that isn't something we're interested in encouraging, but it also doesn't seem worth extra complexity to make slightly harder.

miscco replied:
You are absolutely correct. Thanks.

miscco commented on stl/inc/yvals_core.h:
```diff
@@ -886,6 +889,54 @@
+#define _FIRST_ARGUMENT_TYPE_NAME  _Unnameable_first_argument
+#define _SECOND_ARGUMENT_TYPE_NAME _Unnameable_second_argument
+#define _RESULT_TYPE_NAME          _Unnameable_result
+#endif // !_HAS_DEPRECATED_ADAPTOR_TYPEDEFS
```
Urgh, how did that slip through. I adapted `_Umap_traits` and `_Uset_traits`

miscco commented on stl/inc/memory:
```diff
@@ -1205,10 +1206,12 @@ public:
 
-    _CXX17_DEPRECATE_SHARED_PTR_UNIQUE _NODISCARD bool unique() const noexcept {
+#if _HAS_DEPRECATED_SHARED_PTR_UNIQUE
+    _NODISCARD _CXX17_DEPRECATE_SHARED_PTR_UNIQUE bool unique() const noexcept {
```
yikes, sorry

StephanTLavavej commented:
Working on updating affected tests; no product code issues found.

StephanTLavavej commented:
Thank you for helping encourage the C++ community to modernize their codebases! This has passed our internal testing and has been merged for VS 2019 16.6 Preview 2. We’ll need to report issues in various “Real World Code” projects (e.g. Boost) this week as our extended test suites run.

StephanTLavavej commented:
Also filed [Dolphin-11970](https://bugs.dolphin-emu.org/issues/11970) "fmt 5.3.0 is incompatible with C++20" (fixed in fmt 6.0.0).
